common/metrics/structured/doc.go

```diff
@@ -0,0 +1,119 @@
+/*
+Package structured contains the base objects for a struct-based metrics system.
+
+This is intended to be used with internal/tools/metricsgen, but the Emitter is
+public on many StructTags to ensure ad-hoc metrics are still simple to emit (and
+to make codegen reasonably easy).
+
+For concrete details, check the generated code of any ...Tags structs, or the
+generator in [github.com/uber/cadence/internal/tools/metricsgen].
+
+# To make a new metrics-tag-containing struct
+
+  - Define a `type ...Tags struct` anywhere.  These can be public or private.
+  - Embed any parent ...Tags structs desired, and add any fields to store tag values
+    (or declare that they will be emitted, if they are not static)
+  - Add a `//go:generate metricsgen` comment to the file (if not already present)
+  - Run `make metrics` to generate the supporting code
+
+In many cases, that's likely enough.  Construct your new thing and use it:
+
+	thing := NewYourTags(parents, and, tags) // get it from somewhere
+	thing.Count("name", 1)                   // "name" must be unique within Cadence
+	// or inside a method on YourTags:
+	func (y YourTags) ItHappened() {
+		y.Count("it_happened", 1)
+	}
+
+to emit a metric with all the associated tags.
+
+# To add new tags to existing metrics / structs
+
+Add the field and run `make metrics`.
+
+This will re-generate the constructor(s), which will lead to a broken build.
+Just chase build failures until you've ensured that every code path has access
+to the new data you wanted to add.
+
+# To see what tags an existing metric has
+
+Find the name string (e.g. grep for it), open it in an IDE, and just ask the
+IDE to auto-complete a field access:
+
+	yourTagsInstance.<ctrl-space to request autocomplete>
+
+In Goland, VSCode, and likely elsewhere, this will give you a drop-down of all
+fields inherited from all parents, for easy reading.
+
+# Best practices
+
+Use constant, in-line strings for metric names.  Prometheus requires that each
+"name" must have a stable set of tags, so there is no safety benefit to using a
+const - generally speaking it must NOT be shared.
```

Comment:
this might need rephrasing, I think I know what you mean about putting the strings inline, there being no value in having separate constants, though i'm unclear about the problem with sharing these names?